File: experiments/issue-288/ground_truth.py
"""Print the canonical rendering of every case in issue #288."""
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "python"))
from links_notation import Parser  # noqa: E402

CASES = [
    '(a " " b)', '(a "" b)', "(a '' b)", '(a `` b)',
    '(a "" "" b)', "(a '' '' b)", '(a `` `` b)',
    '(a ""x"" b)', '(a """" b)', '(x "" " "")', '(x \' " \')',
    '("" ("" 1))', '("" (\'\' 1))', '("x" ("" 1))', '("" ("x" 1))',
    '("" x ("" 1))', '("" 1 ("" 1))', '(o: ("" (o: ("" 1))))',
    '(a " b)', '(a """ b)', '("")', '("": 1)', '(a ""  "" b)', '("" "")',
]


def render(node):
    if not node.values:
        return "<%s>" % ("" if node.id is None else node.id)
    head = "" if node.id is None else "<%s>: " % node.id
    return "(%s%s)" % (head, " ".join(render(v) for v in node.values))


def main():
    parser = Parser()
    for case in CASES:
        try:
            links = parser.parse(case)
            print("%-24s => %s" % (case, "\n".join(render(l) for l in links)))
        except Exception as e:  # noqa: BLE001
            print("%-24s => Err(%s)" % (case, e))


if __name__ == "__main__":
    main()

File: experiments/issue-288/java/GroundTruth.java
import io.github.linkfoundation.linksnotation.Link;
import io.github.linkfoundation.linksnotation.Parser;
import java.util.List;
import java.util.stream.Collectors;

/** Print the canonical rendering of every case in issue #288. */
public final class GroundTruth {
  private static final String[] CASES = {
    "(a \" \" b)", "(a \"\" b)", "(a '' b)", "(a `` b)",
    "(a \"\" \"\" b)", "(a '' '' b)", "(a `` `` b)",
    "(a \"\"x\"\" b)", "(a \"\"\"\" b)", "(x \"\" \" \"\")", "(x ' \" ')",
    "(\"\" (\"\" 1))", "(\"\" ('' 1))", "(\"x\" (\"\" 1))", "(\"\" (\"x\" 1))",
    "(\"\" x (\"\" 1))", "(\"\" 1 (\"\" 1))", "(o: (\"\" (o: (\"\" 1))))",
    "(a \" b)", "(a \"\"\" b)", "(\"\")", "(\"\": 1)", "(a \"\"  \"\" b)", "(\"\" \"\")",
  };

  private static String render(Link node) {
    if (node.getValues() == null || node.getValues().isEmpty()) {
      return "<" + (node.getId() == null ? "" : node.getId()) + ">";
    }
    String head = node.getId() == null ? "" : "<" + node.getId() + ">: ";
    return "("
        + head
        + node.getValues().stream().map(GroundTruth::render).collect(Collectors.joining(" "))
        + ")";
  }

  public static void main(String[] args) {
    Parser parser = new Parser();
    for (String source : CASES) {
      try {
        List<Link> links = parser.parse(source);
        System.out.printf(
            "%-24s => %s%n",
            source, links.stream().map(GroundTruth::render).collect(Collectors.joining("\n")));
      } catch (Exception e) {
        System.out.printf("%-24s => Err(%s: %s)%n", source, e.getClass().getSimpleName(), e.getMessage());
      }
    }
  }
}
